pennylane/ops/functions/matrix.py

```diff
@@ -134,13 +134,27 @@ def _matrix(tape, wire_order=None):
     interface = qml.math._multi_dispatch(params)
 
     wire_order = wire_order or tape.wires
-    n_wires = len(wire_order)
 
     # initialize the unitary matrix
-    unitary_matrix = qml.math.eye(2**n_wires, like=interface)
+    unitary_matrix = qml.math.eye(2 ** len(wire_order), like=interface)
 
+    broadcasted_unitary = False
     for op in tape.operations:
         U = matrix(op, wire_order=wire_order)
-        unitary_matrix = qml.math.tensordot(U, unitary_matrix, axes=[[-1], [-2]])
+        broadcasted_U = qml.math.ndim(U) == 3

```

Comment:
Minor: suggest renaming, the name `broadcasted_U` faintly tells me that the object is a broadcasted unitary whereas it's a boolean describing whether or not `U` was broadcasted.
```suggestion
        U_is_broadcasted = qml.math.ndim(U) == 3
```